x/distribution/src/keeper/validator.rs

```diff
@@ -0,0 +1,170 @@
+use super::*;
+use crate::{ValidatorCurrentRewards, ValidatorHistoricalRewards};
+use gears::{
+    error::AppError,
+    types::{decimal256::Decimal256, uint::Uint256},
+};
+
+impl<
+        SK: StoreKey,
+        PSK: ParamsSubspaceKey,
+        AK: AuthKeeper<SK, M>,
+        BK: BankKeeper<SK, M>,
+        DSK: DistributionStakingKeeper<SK, M>,
+        M: Module,
+    > Keeper<SK, PSK, AK, BK, DSK, M>
+{
+    /// increment the reference count for a historical rewards value
+    pub fn increment_reference_count<DB: Database>(
+        &self,
+        ctx: &mut TxContext<'_, DB, SK>,
+        validator_address: &ValAddress,
+        period: u64,
+    ) -> Result<(), AppError> {
+        let mut historical = self
+            .validator_historical_rewards(ctx, validator_address, period)?
+            .ok_or(AppError::Custom(
+                "historical rewards are not found".to_string(),
+            ))?;
+        if historical.reference_count > 2 {
+            // TODO: sdk behaviour, seems to be correct
+            panic!("reference count should never exceed 2")
```

Comment:
This is a bit worrying. Do you expect it not to happen? Can we capture this in a type?